transmission/transmission.go

```diff
@@ -2,80 +2,157 @@ package transmission
+		return
+	}
 }
 
-func InitLibhoney(writeKey string, apiHost string) error {
+func InitLibhoney(apiKey string, apiHost string, bufferSize int, bufferExpire time.Duration) error {
 	err := libhoney.Init(libhoney.Config{
-		WriteKey: writeKey,
-		APIHost:  apiHost,
+		APIKey:      apiKey,
+		APIHost:     apiHost,
+		BlockOnSend: true,
 	})
 	if err != nil {
 		return err
 	}
+
+	ringBuffer = NewRingBuffer(bufferSize, bufferExpire)
+
 	go readResponses()
 	return nil
 }
 
-// TODO: support retrying, etc.
 func readResponses() {
-	for resp := range libhoney.Responses() {
-		// If an event is too large, include the first part of the raw message
+	for resp := range libhoney.TxResponses() {
 		if resp.Err != nil && strings.HasPrefix(resp.Err.Error(), "event exceeds max event size") {
-			rawMessage, ok := resp.Metadata.(string)
-			// if this is somehow not a string, handle the error normally below
+			// error sending event due to size, try to find it in cache
+			key, ok := resp.Metadata.(uint64)
+			if ok {
+				ev, exists := ringBuffer.Get(key)
+
+				if exists {
+					rawMessage := ev.RawMessage
+
+					// include the first part of the raw message
+					msgLen := len(rawMessage)
+					truncatedMsgLen := truncatedLineMax
+					// This shouldn't happen - if the event is too large for the API, it's at least
+					// 1000 bytes, but be defensive here anyway
+					if msgLen < truncatedMsgLen {
+						truncatedMsgLen = msgLen
+					}
+
+					logrus.WithFields(logrus.Fields{
+						"error":             resp.Err,
+						"status":            resp.StatusCode,
+						"responseBody":      string(resp.Body),
+						"truncated_raw_msg": rawMessage[:truncatedMsgLen],
+						"raw_msg_len":       msgLen,
+					}).Error("Unable to send event to Honeycomb API due to size")
+
+					continue
+				}
+			}
+
+			// if we get here we couldn't find the response metadata or event in send buffer
+			logrus.WithFields(logrus.Fields{
+				"error":        resp.Err,
+				"status":       resp.StatusCode,
+				"responseBody": string(resp.Body),
+			}).Error("Unable to send event to Honeycomb API due to size. Event not found in local send buffer.")
+
+		} else if resp.Err != nil || (resp.StatusCode != 200 && resp.StatusCode != 202) {
+			// error sending event, try to find it in buffer
+			key, ok := resp.Metadata.(uint64)
 			if ok {
-				msgLen := len(rawMessage)
-				truncatedMsgLen := truncatedLineMax
-				// This shouldn't happen - if the event is too large for the API, it's at least
-				// 1000 bytes, but be defensive here anyway
-				if msgLen < truncatedMsgLen {
-					truncatedMsgLen = msgLen
+				ev, exists := ringBuffer.Get(key)
+
+				if exists {
+					// event found, we can retry it
+					logrus.WithFields(logrus.Fields{
+						"error":        resp.Err,
+						"status":       resp.StatusCode,
+						"responseBody": string(resp.Body),
+					}).Error("Failed to send event to Honeycomb. Will retry.")
+
+					// Resend event
+					ht := &HoneycombTransmitter{}
```

Comment:
Do we need to create a new transmitter for each retry attempt?

Reply:
We need to put the event back on the ring buffer to retry, so we need to call `Send` on a transmitter. HoneycombTrasmistter is an empty struct, so this shouldn't have any impact.